### ISSUE TITLE
Test persistent_mapping_forget_then_invalidate failing

### ISSUE THREAD

**Issue by nicokoch** [tomaka/glium#1290]
Running the tests on my windows box (Intel GC) and the test is failing with the following output:
(I am on rust nightly btw)

```
failures:

---- persistent_mapping_forget_then_invalidate stdout ----
        Debug message with high or medium severity: `Error has been generated. GL error GL_INVALID_OPERATION in InvalidateBufferData: (ID: 901315929) Generic error`.
Please report this error: https://github.com/tomaka/glium/issues
Backtrace:
   1 - 0x72d29e
   2 - 0x72d157
   3 - 0x6a26e9
   4 - 0x6a2105
   5 - 0x7ffaade4082e
         RegisterProcTableCallback at <unknown>:??
   6 - 0x7ffaade40a57
         RegisterProcTableCallback at <unknown>:??
   7 - 0x7ffaade40b44
         RegisterProcTableCallback at <unknown>:??
   8 - 0x7ffaade430ab
         RegisterProcTableCallback at <unknown>:??
   9 - 0x7ffaadebc1b2
         RegisterProcTableCallback at <unknown>:??
  10 - 0x56fb3a
  11 - 0x56f911
  12 - 0x47e632
  13 - 0x4ccdc7
  14 - 0x4fc969
  15 - 0x4ff0f3
  16 - 0x4fd269
  17 - 0xadfb02
  18 - 0x4fd404
  19 - 0xaf4c3e
  20 - 0x7ffad9fc2d92
         BaseThreadInitThunk at <unknown>:??

thread 'persistent_mapping_forget_then_invalidate' panicked at 'GL_INVALID_OPERATION', src\context/mod.rs:478


failures:
    persistent_mapping_forget_then_invalidate
```

**Comment by tomaka** [comment]
> GL_INVALID_OPERATION is generated if any part of buffer is currently mapped.

This is exactly the purpose of the test: make sure that the buffer is automatically being unmapped before `glInvalidateBufferData` is called.

This means that it's weird that the test is failing for you and not for me. It shouldn't be something platform-specific.

**Comment by tomaka** [comment]
Since I have no idea where this is coming from, could you replace `StructGenerator` with `DebugStructGenerator` [here](https://github.com/tomaka/glium/blob/1eeb32209e19cd07d27bf6a0fdb34c1977df9408/build/main.rs#L146) and run the test again?

It should print the list of all OpenGL functions that are executed.

It will also make the test succeed, so you need to run `cargo test --test buffer -- persistent_mapping_forget_then_invalidate --nocapture`

**Comment by tomaka** [comment]
Please ignore my previous comments. I didn't realize this was persistent mapping.
Will fix it.

**Comment by nicokoch** [comment]
#1293 did not fix this for me, you should consider reopening.

Can I do anything to help debug?

**Comment by tomaka** [comment]
Please try using the `DebugStructGenerator` I explained above.

**Comment by nicokoch** [comment]
The full output (sorry lots of noise) [here](https://gist.github.com/nicokoch/7fe9d68bcdfed553fcd3)

**Comment by tomaka** [comment]
Thanks!

Thanks to this I confirmed that I came from the `invalidate()` function. I took another look at it and fixed the very stupid bug: #1296
